### PR TITLE
feat: scope login attempts to specific subdomains if available - do not switch subdomains

### PR DESCRIPTION
**Pull request by thenanyu**
Functionality
- if a user tries to authenticate from a specific custom domain or team subdomain, the scope of the auth attempt will be limited to that team only.
- even if they have an existing account with their SSO on a different team, we will not consider accounts outside of the team referenced by the url
- if a user authenticates from the apex domain, or if subdomains are not enabled, then it will fall back to the best-effort path, and behave as before

Basic QA test cases
1. **Attempt to auth into a subdomain with an SSO account that exists on a different subdomain. It should fail. Previously, it would have redirected to the other subdomain**
2. Attempt to auth into a subdomain with an SSO account from apex. It should find a team if it exists.
3. Attempt to auth into a subdomain with an SSO that exists on that team. It should succeed.

### CODE REVIEW

File: server/routes/auth/providers/google.ts [hunk 1]
```diff
@@ -63,27 +63,48 @@ if (env.GOOGLE_CLIENT_ID && env.GOOGLE_CLIENT_SECRET) {
...
             // the query parameter.
             const avatarUrl = profile.picture.replace("=s96-c", "=s128-c");
 
+            // if a team can be inferred, we assume the user is only interested in signing into
+            // that team in particular; otherwise, we will do a best effort at finding their account
+            // or provisioning a new one (within AccountProvisioner)
             result = await accountProvisioner({
               ip: req.ip,
               team: {
+                id: team?.id,
```
Comment (thenanyu) [inline server/routes/auth/providers/google.ts:107]: 🥦 team id is passed into the account provisioner if available, and will be respected as the only team which the command chain will consider for login/signup

File: server/commands/teamCreator.ts [hunk 2]
```diff
@@ -33,7 +35,9 @@ async function teamCreator({
   ip,
 }: Props): Promise<TeamCreatorResult> {
   let authP = await AuthenticationProvider.findOne({
-    where: authenticationProvider,
+    where: id
+      ? { ...authenticationProvider, teamId: id }
+      : authenticationProvider,
```
Comment (thenanyu) [inline server/commands/teamCreator.ts:40]: 🥦 2/ do a best effort here to find the right team. If a `teamId` is provided, we're interpreting that as the only team we're interested in.

File: server/commands/userCreator.ts [hunk 3]
```diff
@@ -48,6 +48,8 @@ export default async function userCreator({
       {
         model: User,
         as: "user",
+        where: { teamId },
+        required: true,
```
Comment (thenanyu) [inline server/commands/userCreator.ts:52]: 🥦 3/ only consider users from the target `teamId`. This opens the door to having the same external `providerId` on multiple users on multipole teams

File: server/commands/teamCreator.ts [hunk 4]
```diff
@@ -33,7 +35,9 @@ async function teamCreator({
   ip,
 }: Props): Promise<TeamCreatorResult> {
   let authP = await AuthenticationProvider.findOne({
-    where: authenticationProvider,
+    where: id
+      ? { ...authenticationProvider, teamId: id }
```
Comment (tommoor) [inline server/commands/teamCreator.ts:39]: Note there is currently no index on `teamId` – need to add this now we're querying it

File: server/routes/auth/providers/oidc.ts [hunk 5]
```diff
@@ -85,10 +90,13 @@ if (env.OIDC_CLIENT_ID && env.OIDC_CLIENT_SECRET) {
             throw OIDCMalformedUserInfoError();
           }
 
-          const subdomain = domain.split(".")[0];
+          // remove the TLD and form a subdomain from the remaining
+          const subdomain = domain.split(".").slice(0, -1).join("-");
```
Comment (tommoor) [inline server/routes/auth/providers/oidc.ts:94]: Wanna move this to a utils method too? Seems like it changed in a couple of spots

File: server/routes/auth/providers/oidc.ts [hunk 6]
```diff
@@ -1,5 +1,5 @@
 import passport from "@outlinewiki/koa-passport";
-import { Request } from "koa";
+import type { Request } from "express";
```
[the remark this answers is not in the document]
Comment (tommoor) [reply server/routes/auth/providers/oidc.ts:2]: Yea, ideally we need to overwrite the type at the global level – it's definitely _not_ an Express request, express doesn't even exist in the lockfile – just it's types.

Then the question is why isn't `koa.Request` including the cookies property if it's actually there

**Review by tommoor — COMMENTED** [review]
This is great

File: server/commands/teamCreator.ts (hunk 4 above)
Comment (thenanyu) [reply server/commands/teamCreator.ts:43]: I don't think performance is any worse than before — the addition of teamId should be applied after the query planner makes use of existing indices

Comment (tommoor) [reply server/commands/teamCreator.ts:43]: I suppose it still has the providerId, ok 👍

File: server/errors.ts [hunk 7]
```diff
@@ -1,16 +1,25 @@
 import httpErrors from "http-errors";
-import env from "./env";
 
 export function AuthenticationError(
-  message = "Invalid authentication",
-  redirectUrl = env.URL
+  message = "Authentication required",
+  redirectUrl = "/"
```
Comment (thenanyu) [inline server/errors.ts:5]: changed these to be relative to preserve subdomain and custom domains

File: server/routes/auth/providers/azure.ts [hunk 8]
```diff
@@ -36,7 +40,7 @@ if (env.AZURE_CLIENT_ID && env.AZURE_CLIENT_SECRET) {
       scope: scopes,
     },
     async function (
-      req: Request,
+      req: Context,
```
[the remark this answers is not in the document]
Comment (tommoor) [reply server/routes/auth/providers/azure.ts:43]: Ultimately I guess we should rename `req` to `ctx` to clarify then

File: server/routes/auth/providers/azure.ts [hunk 9]
```diff
@@ -36,7 +40,7 @@ if (env.AZURE_CLIENT_ID && env.AZURE_CLIENT_SECRET) {
       scope: scopes,
     },
     async function (
-      req: Request,
+      ctx: Context,
```
Comment (thenanyu) [inline server/routes/auth/providers/azure.ts:43]: changed referenced from "request" to "context"

**Review by tommoor — APPROVED** [review]

File: server/commands/teamCreator.ts (hunk 2 above)
Comment (smotraghi-shopify) [reply server/commands/teamCreator.ts:44]: Sounds good - will do

**Pull request merged**